extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,518 @@
+
+        @Override
+        public String getServiceName() {
+            return getClass().getSimpleName();
+        }
+
+        /**
+         * Core message pulling logic code goes here.
+         */
+        private void doPull() {
+            for (DruidPullRequest pullRequest : requestsRead) {
+                PullResult pullResult;
+                try {
+                    if (!pullRequest.isLongPull()) {
+                        pullResult = defaultMQPullConsumer.pull(
+                                pullRequest.getMessageQueue(),
+                                pullRequest.getTag(),
+                                pullRequest.getNextBeginOffset(),
+                                pullRequest.getPullBatchSize());
+                    } else {
+                        pullResult = defaultMQPullConsumer.pullBlockIfNotFound(
+                                pullRequest.getMessageQueue(),
+                                pullRequest.getTag(),
+                                pullRequest.getNextBeginOffset(),
+                                pullRequest.getPullBatchSize()
+                        );
+                    }
+
+                    switch(pullResult.getPullStatus()) {
+                        case FOUND:
+                            // Handle pull result.
+                            if (!messageQueueTreeSetMap.keySet().contains(pullRequest.getMessageQueue())) {
+                                messageQueueTreeSetMap.putIfAbsent(pullRequest.getMessageQueue(),
+                                        new ConcurrentSkipListSet<>(new MessageComparator()));
+                            }
+                            messageQueueTreeSetMap.get(pullRequest.getMessageQueue()).addAll(pullResult.getMsgFoundList());
+                            break;
+
+                        case NO_NEW_MSG:
+                        case NO_MATCHED_MSG:
+                            break;
+
+                        case OFFSET_ILLEGAL:
+                            LOGGER.error("Bad Pull Request: Offset is illegal. Offset used: {}",
+                                    pullRequest.getNextBeginOffset());
+                            break;
+
+                        default:
+                            break;
+                    }
+                } catch (MQClientException | RemotingException | MQBrokerException | InterruptedException e) {
+                    LOGGER.error("Failed to pull message from broker.", e);
+                } finally {
+                    pullRequest.getCountDownLatch().countDown();
+                }
+
+            }
+            requestsRead.clear();
+        }
+
+        /**
+         * Thread looping entry.
+         */
+        @Override
+        public void run() {
+            LOGGER.info(getServiceName() + " starts.");
+            while (!isStoped()) {
+                waitForRunning(0);
+                doPull();
+            }
+
+            // in case this service is shutdown gracefully without interruption.
```

Comment:
This is to finish the last round completely, aka, all requests received are processed in view of DruidPullMessageService. It does not bear much significance though.